File: src/main/java/com/lessmarkup/interfaces/recordmodel/InputFieldDefinition.java
package com.lessmarkup.interfaces.recordmodel;

import com.lessmarkup.framework.helpers.PropertyDescriptor;
import com.lessmarkup.framework.helpers.StringHelper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class InputFieldDefinition {

    private final PropertyDescriptor property;
    private final String id;
    private final boolean readOnly;
    private final InputFieldType type;
    private final String readOnlyCondition;
    private final boolean required;
    private final String textId;
    private final String visibleCondition;
    private final double width;
    private final int minWidth;
    private final int maxWidth;
    private final int position;
    private final String defaultValue;
    private List<InputFieldEnum> enumValues;
    
    public InputFieldDefinition(PropertyDescriptor property, InputField definition) {
        this.property = property;
        if (StringHelper.isNullOrEmpty(definition.id())) {
            this.id = StringHelper.toJsonCase(property.getName());
        } else {
            this.id = definition.id();
        }
        this.readOnly = definition.readOnly();
        this.readOnlyCondition = definition.readOnlyCondition();
        this.required = definition.required();
        this.textId = definition.textId();
        this.type = definition.type();
        this.visibleCondition = definition.visibleCondition();
        this.width = definition.width();
        this.minWidth = definition.minWidth();
        this.maxWidth = definition.maxWidth();
        this.position = definition.position();
        this.defaultValue = definition.defaultValue();

        if (property.getType().isEnum() &&
                (type == InputFieldType.SELECT || type == InputFieldType.SELECT_TEXT || type == InputFieldType.MULTI_SELECT) && 
                definition.enumTextIdBase() != null) {
            initializeEnum(property.getType(), definition.enumTextIdBase());
        }
    }
    
    private void initializeEnum(Class<?> enumType, String textIdBase) {
        
        this.enumValues = new ArrayList<>();
        
        Arrays.stream(enumType.getEnumConstants()).forEach(c -> {
            String enumTextId = textIdBase + c.toString();
            InputFieldEnum enumValue = new InputFieldEnum(enumTextId, c.toString());
            this.enumValues.add(enumValue);
        });
    }
    
    public String getId() { return id; }
    public PropertyDescriptor getProperty() { return property; }
    public boolean isRequired() { return required; }
    public InputFieldType getType() { return type; }
    public boolean isReadOnly() { return readOnly; }
    public String getVisibleCondition() { return visibleCondition; }
    public String getReadOnlyCondition() { return readOnlyCondition; }
    public String getTextId() { return textId; }
    public double getWidth() { return width; }
    public int getMinWidth() { return minWidth; }
    public int getMaxWidth() { return maxWidth; }
    public int getPosition() { return position; }
    public String getDefaultValue() { return defaultValue; }
    public List<InputFieldEnum> getEnumValues() { return enumValues; }
}

File: src/main/scala/com/lessmarkup/interfaces/recordmodel/InputFieldDefinition.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.recordmodel

import com.lessmarkup.framework.helpers.{PropertyDescriptor, StringHelper}
import com.lessmarkup.interfaces.annotations
import com.lessmarkup.interfaces.annotations.{InputFieldType, InputField}

class InputFieldDefinition(property: PropertyDescriptor, definition: InputField) {
  private final val id: String =
    if (StringHelper.isNullOrEmpty(definition.id)) StringHelper.toJsonCase(property.getName)
    else definition.id

  private final val readOnly: Boolean = definition.readOnly()
  private final val `type`: annotations.InputFieldType = definition.fieldType()
  private final val readOnlyCondition: String = definition.readOnlyCondition()
  private final val required: Boolean = definition.required()
  private final val textId: String = definition.textId()
  private final val visibleCondition: String = definition.visibleCondition()
  private final val width: Double = definition.width()
  private final val minWidth: Int = definition.minWidth()
  private final val maxWidth: Int = definition.maxWidth()
  private final val position: Int = definition.position()
  private final val defaultValue: String = definition.defaultValue()
  private final val enumValues: List[InputFieldEnum] = initializeEnum()
  private final val inlineWithPrevious: Boolean = definition.inlineWithPrevious()

  private def initializeEnum(): List[InputFieldEnum] = {
    if (!(property.getType.isEnum && (`type` == InputFieldType.SELECT || `type` == InputFieldType.SELECT_TEXT || `type` == InputFieldType.MULTI_SELECT)) || definition.enumTextIdBase == null) {
      return List()
    }
    val enumType = property.getType
    val textIdBase = definition.enumTextIdBase
    enumType.getEnumConstants.map(c => {
      val enumTextId = textIdBase + c.toString
      new InputFieldEnum(enumTextId, c.toString)
    })
    .toList
  }

  def getId: String = id

  def getProperty: PropertyDescriptor = property

  def isRequired: Boolean = required

  def getType: annotations.InputFieldType = `type`

  def isReadOnly: Boolean = readOnly

  def getVisibleCondition: String = visibleCondition

  def getReadOnlyCondition: String = readOnlyCondition

  def getTextId: String = textId

  def getWidth: Double = width

  def getMinWidth: Int = minWidth

  def getMaxWidth: Int = maxWidth

  def getPosition: Int = position

  def getDefaultValue: String = defaultValue

  def getEnumValues: List[InputFieldEnum] = enumValues

  def isInlineWithPrevious: Boolean = inlineWithPrevious
}